plutus-scb/src/Cardano/Protocol.hs

```diff
@@ -0,0 +1,212 @@
+
+module Cardano.Protocol where
+
+import           Codec.Serialise.Class           (Serialise)
+import           Data.Aeson                      (FromJSON, ToJSON)
+import qualified Data.ByteString.Lazy            as BSL
+import           Data.Functor                    (void)
+import           Data.List                       (findIndex)
+import           Data.Text.Prettyprint.Doc       (Pretty)
+import           Eventful.Projection             (GlobalStreamProjection, Projection (..), StreamProjection (..))
+import           Eventful.Store.Class
+import           GHC.Generics
+
+import           Control.Lens                    hiding (index, modifying, use)
+import           Control.Monad.Freer             (Eff, Member, Members, type (~>), interpret, send)
+import           Control.Monad.Freer.Extra.Log
+import           Control.Monad.Freer.Extra.State
+import           Control.Monad.Freer.State       as Eff
+import           Control.Monad.Freer.Writer
+
+import           Cardano.Prelude                 (NoUnexpectedThunks)
+import           Cardano.Slotting.Slot           (SlotNo (..), WithOrigin (..))
+import           Ouroboros.Network.Block         (HeaderHash, Point (..), StandardHash, pointSlot)
+import qualified Ouroboros.Network.Point         as Point
+
+import           Ledger                          (Block, Slot (..), Tx (..), TxId (..), txId)
+import qualified Ledger.Index                    as Index
+import           LedgerBytes                     (LedgerBytes (..))
+import           Plutus.SCB.EventLog
+import           Plutus.SCB.Events
+import           Plutus.SCB.Utils                (tshow)
+import           Wallet.Emulator.Chain           (ChainEffect (..), ChainState (..), chainNewestFirst, currentSlot,
+                                                  emptyChainState, handleChain, index, processBlock, queueTx)
+import qualified Wallet.Emulator.Chain           as EM
+import           Wallet.Emulator.NodeClient      (BlockValidated (..), NodeClientEffect (..), NodeClientEvent (..),
+                                                  NodeClientState (..), handleNodeClient)
+
+-- Block header
+newtype BlockId = BlockId { getBlockId :: BSL.ByteString }
+  deriving (Eq, Ord, Generic)
+  deriving anyclass (ToJSON, FromJSON)
+  deriving newtype (Serialise, NoUnexpectedThunks)
+  deriving (Pretty, Show) via LedgerBytes
+
+blockId :: Block -> BlockId
+blockId = BlockId . foldl BSL.append BSL.empty . map (getTxId . txId)
+
+-- Making txs work with the node protocols
+deriving instance StandardHash Tx
+deriving instance StandardHash Block
+deriving newtype instance NoUnexpectedThunks TxId
+type instance HeaderHash Tx = TxId
+type instance HeaderHash Block = BlockId
+
+type ClientStreamProjection = GlobalStreamProjection ChainState ChainEvent
+
+chainState :: Lens' ClientStreamProjection ChainState
+chainState = lens getter setter where
+    getter = streamProjectionState
+    setter prj cs = prj { streamProjectionState = cs }
```

Comment:
Is this safe? Shouldn't we be using the eventful functions to update this rather than setting it ourselves?